numpy/lib/_function_base_impl.py

```diff
@@ -1867,62 +1867,130 @@ def sort_complex(a):
         return b
 
 
-def _trim_zeros(filt, trim=None):
+def _arg_trim_zeros(filt):
+    """Return indices of the first and last non-zero element.
+
+    Parameters
+    ----------
+    filt : array_like
+        Input array.
+
+    Returns
+    -------
+    start, stop : ndarray
+        Two arrays containing the indices of the first and last non-zero
+        element in each dimension.
+
+    See also
+    --------
+    trim_zeros
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> _arg_trim_zeros(np.array([0, 0, 1, 1, 0]))
+    (array([2]), array([3]))
+    """
+    nonzero = np.argwhere(filt != 0)
+    if nonzero.size == 0:
+        start = stop = np.array([], dtype=np.intp)
+    else:
+        start = nonzero.min(axis=0)
+        stop = nonzero.max(axis=0)
+    return start, stop
+
+
+def _trim_zeros(filt, trim=None, axis=None):
     return (filt,)
 
 
 @array_function_dispatch(_trim_zeros)
-def trim_zeros(filt, trim='fb'):
-    """
-    Trim the leading and/or trailing zeros from a 1-D array or sequence.
+def trim_zeros(filt, trim='fb', axis=None):
+    """Remove values along a dimension which are zero along all other.
 
     Parameters
     ----------
-    filt : 1-D array or sequence
+    filt : array_like
         Input array.
-    trim : str, optional
+    trim : {"fb", "f", "b"}, optional
         A string with 'f' representing trim from front and 'b' to trim from
-        back. Default is 'fb', trim zeros from both front and back of the
-        array.
+        back. By default, zeros are trimmed from the front and back.
+    axis : int or sequence, optional
+        The axis to trim. If None, the default, all axes are trimmed.
 
     Returns
     -------
-    trimmed : 1-D array or sequence
+    trimmed : ndarray or sequence
         The result of trimming the input. The input data type is preserved.
 
+    Notes
+    -----
+    For all-zero arrays, the first axis is trimmed first.
```

Comment:
Also note how this function can be applied for values other than zero.

Reply:
After playing around with masked arrays. Even if I do something about the internal call to `np.asarray` that converts the masked array back

```python
import numpy as np
b = np.array([[1, 1, 2, 3, 1, 1],
              [1, 1, 0, 3, 0, 1],
              [1, 1, 1, 1, 1, 1]])
b_masked = np.ma.MaskedArray(b, mask=(b == 1))
np.trim_zeros(b_masked)
```
will still trim zeros as well, in this case the zero at `b[1, 4]`. It shouldn't in a case where we only want to trim ones.

The alternative I see is to make the currently internal function `_arg_trim_zeros` public. But might be another PR.

Reply:
I am not sure where the masked array discussion comes from?  I can see a `mask=` being passed in, in principle (to be used instead of `values !=/== 0`), but we shouldn't worry about masked arrays unless it is to not break what currently happens to work.

Reply:
That was a concern raised during the triage meeting. The question was along the lines of "Why is zero special? Why shouldn't this work for other values?". 

The suggestion was to explore solutions with masked arrays and document it as an example.

Reply:
Maybe just add a note about the behavior with masked arrays in the docstring?